### PR TITLE
C++: Improve UseOfStringAfterLifetimeEnds doc.

### PR DESCRIPTION
**Pull request by geoffw0**
Improve `cpp/use-of-string-after-lifetime-ends` `.qhelp`, references and alert message.  The aim is to be more helpful for users who are not very familiar with temporary objects and lifetime issues that can result.

@MathiasVP please check for technical inaccuracies.

Docs team, please make any suggestions you may have for readability etc.

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 1]
```diff
@@ -8,6 +8,12 @@
 When the <code>std::string</code> object is destroyed, the pointer returned by <code>c_str</code> is no
 longer valid. If the pointer is used after the <code>std::string</code> object is destroyed, then the behavior is undefined.
 </p>
+
+<p>Typically this problem occurs when a <code>std::string</code> is returned by a function call (or overloaded operator)
+by value, and the result is not immediately stored in a variable by value (or <code>const</code> reference). The resulting
```
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:13]: There's a third option as well (which is probably the most idiomatic way to fix this nowadays): Storing the temporary string in a rvalue reference. For example:
```cpp
std::string&& s = std::string("abc");
```
The benefit of this (as opposed to storing it in a `const std::string& s`) is that the variable remains mutable.

Maybe it's better to say something like:
> Typically this problem occurs when a <code>std::string</code> is returned by a function call (or overloaded operator)
by value, and the result is not immediately stored in a variable in a way that extends the lifetime of the temporary object

and then we can maybe add a link to https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary?

Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp]: Looks good. I'm happy with this!

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 2]
```diff
@@ -8,6 +8,12 @@
 When the <code>std::string</code> object is destroyed, the pointer returned by <code>c_str</code> is no
 longer valid. If the pointer is used after the <code>std::string</code> object is destroyed, then the behavior is undefined.
 </p>
+
+<p>Typically this problem occurs when a <code>std::string</code> is returned by a function call (or overloaded operator)
```
Comment (mchammer01) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:12]: Minor
```suggestion
<p>Typically, this problem occurs when a <code>std::string</code> is returned by a function call (or overloaded operator)
```

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 3]
```diff
@@ -8,6 +8,12 @@
...
+by value, and the result is not immediately stored in a variable by value or reference in a way that extends the lifetime of
+the temporary object. The resulting temporary <code>std::string</code> object is destroyed at the end of the expression
```
Comment (mchammer01) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:14]: What about this suggestion to make this sentence shorter? 🤔 
Feel free to ignore if you don't agree, or if it doesn't work here 😅 
```suggestion
the temporary object. The resulting temporary <code>std::string</code> object is destroyed at the end of the containing expression
```

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 4]
```diff
@@ -8,6 +8,12 @@
...
+statement it is contained in, along with any memory returned by a call to <code>c_str</code>.
```
Comment (mchammer01) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:15]: Part 2 of the suggestion above.
```suggestion
statement, along with any memory returned by a call to <code>c_str</code>.
```

**Review by mchammer01 — DISMISSED** [review]
@geoffw0 - LGTM from a Docs perspective ✨ 
Made a few minor comment/suggestions.

**Comment by geoffw0** [comment]
Suggestions accepted, thank you @mchammer01 !

**Review by mchammer01 — APPROVED** [review]
LGTM :shipit:

**Pull request merged**